v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/writer/DeadLetterQueue.java

```diff
@@ -0,0 +1,146 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.writer;
+
+import com.google.cloud.spanner.Mutation;
+import com.google.cloud.teleport.v2.constants.MetricCounters;
+import com.google.cloud.teleport.v2.templates.RowContext;
+import com.google.cloud.teleport.v2.transforms.DLQWriteTransform;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.Serializable;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.io.FileSystems;
+import org.apache.beam.sdk.io.gcp.spanner.MutationGroup;
+import org.apache.beam.sdk.metrics.Counter;
+import org.apache.beam.sdk.metrics.Metrics;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+import org.checkerframework.checker.initialization.qual.Initialized;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.UnknownKeyFor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Manages the dead letter queue in the pipeline. */
+public class DeadLetterQueue implements Serializable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DeadLetterQueue.class);
+
+  private final String dlqDirectory;
+
+  private final PTransform<PCollection<String>, PDone> dlqTransform;
+
+  public static final Counter failedMutationCounter =
+      Metrics.counter(SpannerWriter.class, MetricCounters.FAILED_MUTATION_ERRORS);
+
+  public static DeadLetterQueue create(String dlqDirectory) {
+    return new DeadLetterQueue(dlqDirectory);
+  }
+
+  public String getDlqDirectory() {
+    return dlqDirectory;
+  }
+
+  public PTransform<PCollection<String>, PDone> getDlqTransform() {
+    return dlqTransform;
+  }
+
+  private DeadLetterQueue(String dlqDirectory) {
+    this.dlqDirectory = dlqDirectory;
+    this.dlqTransform = createDLQTransform(dlqDirectory);
+  }
+
+  @VisibleForTesting
+  private PTransform<PCollection<String>, PDone> createDLQTransform(String dlqDirectory) {
+    if (dlqDirectory != null) {
+      String dlqUri = FileSystems.matchNewResource(dlqDirectory, true).toString();
+      LOG.info("setting up dead letter queue directory: {}", dlqDirectory);
+      return DLQWriteTransform.WriteDLQ.newBuilder()
+          .withDlqDirectory(dlqUri)
+          .withTmpDirectory(dlqUri + "/tmp")
+          .setIncludePaneInfo(true)
+          .build();
+    } else {
+      LOG.warn("writing errors to log as no DLQ directory configured");
+      return new WriteToLog();
+    }
+  }
+
+  public static class WriteToLog extends PTransform<PCollection<String>, PDone> {
```

Comment:
I wouldn't recommend this in general. We should not write individual events to Logs since this can be really expensive and throttle the pipeline due to Cloud logging limitations.

Reply:
Ignoring the errors is arguably worse, hence this is provided as a fallback. We've not seen pipeline throttling owing to this in the large scale tests.